src/ts/content/extensions/util/onLogged.ts

```diff
@@ -44,40 +47,67 @@ const removeInjectedButton = (button: HTMLTableCellElement) => {
 	button.remove();
 };
 
-const handleLog = (text: string, toastType: ToastType, callbacks: OnLoggedCallback[]) => () => {
+const handleStatusChange = (text: string, toastType: ToastType) => async () => {
 	showToast(text, toastType);
-	callbacks.map((callback) => callback());
+	const status = await getLogStatus();
+	if (status.ready) {
+		loginCallbacks.forEach((callback) => callback(status));
+	} else {
+		logoutCallbacks.forEach((callback) => callback(status));
+	}
 };
 
-const handleLogin = handleLog("Logged in!", ToastType.SUCCESS, loginCallbacks);
-const handleLogout = handleLog("Logged out!", ToastType.INFO, logoutCallbacks);
-
-const onClick = () => loaderOverlaid(() => authorize(true).catch(console.error));
+const onLoginClick = () => loaderOverlaid(() => authorize(true).catch(console.error));
+const onAddSheetClick = () => invokeWorker(WorkerKind.OpenOptions, null);
 
 const compose =
-	(...listeners: OnLoggedCallback[]): OnLoggedCallback =>
-	() =>
-		listeners.forEach((listener) => listener?.());
-
-const saveCallback = (callback: OnLoggedCallback, callbacks: OnLoggedCallback[], applyNow: boolean): void => {
+	(...listeners: InternalOnLoggedCallback[]): InternalOnLoggedCallback =>
+	(status) =>
+		listeners.forEach((listener) => listener?.(status));
+
+const saveCallback = (
+	callback: InternalOnLoggedCallback,
+	callbacks: InternalOnLoggedCallback[],
+	status: LoggedStatus,
+	applyNow: boolean
+): void => {
 	callbacks.push(callback);
-	applyNow && callback();
+	applyNow && callback(status);
+};
+
+const getLogStatus = async (): Promise<LoggedStatus> => {
+	const futureAuth = isAuthorized();
+	const futureSheetSet = isSheetSet();
+	const [authed, sheetSet] = await Promise.all([futureAuth, futureSheetSet]);
+	return {authed, sheetSet, ready: authed && sheetSet};
 };
 
-const onLogged = async ({onLogIn, onLogOut, container, description}: OnLogOptions) => {
-	const authed = await isAuthorized();
+const onLogged = async ({onLogIn: userOnLogIn, onLogOut: userOnLogOut, container, description}: OnLogOptions) => {
+	let onLogIn: InternalOnLoggedCallback = userOnLogIn;
+	let onLogOut: InternalOnLoggedCallback = userOnLogOut;
 	if (container) {
-		const button = createIconButton("Login", "img/login.png", onClick, description);
-		onLogIn = compose(() => removeInjectedButton(button), onLogIn);
-		onLogOut = compose(() => injectButton(button, container), onLogOut);
+		const loginButton = createIconButton("Login", "img/login.png", onLoginClick, description);
+		const sheetLinkButton = createIconButton("Add Tag Index", "img/icon16.png", onAddSheetClick);
+		const removeButtons = () => {
+			removeInjectedButton(loginButton);
+			removeInjectedButton(sheetLinkButton);
+		};
+		onLogIn = compose(removeButtons, onLogIn);
```

Comment:
i'm confuzzled, why does this not type error? `onLogIn` is being changed from `OnLoggedCallback` (which doesn't take a `status` param) to `InternalOnLoggedCallback` (which requires it). same thing with `removeButtons`, it doesn't take a `status` either, but is being treated like it does in `compose`

is it because you're optional chaining it?

```ts
const compose =
	(...listeners: InternalOnLoggedCallback[]): InternalOnLoggedCallback =>
	(status) =>
		listeners.forEach((listener) => listener?.(status));
```

Reply:
The optional chaining probably shouldn't actually be there -- it's unrelated.

`OnLoggedCallback ` is a sub-type of `InternalOnLoggedCallback`. The client will supply `status`, and the function may or may not actually use it.

Remember the Liskov substitution principle or whatever for a subtype: widening preconditions (a specific `status` to literally anything getting passed because it's unused), and narrowing postconditions (unchanged)

All functions are a subtype of `(args: ...never[]) => unknown`